pkg/http/server.go

```diff
@@ -0,0 +1,233 @@
+		// No HTTP listener required
+		return nil
+	}
+
+	networkType := getNetworkScheme(opts.BindAddress)
+	listenAddr := getListenAddress(opts.BindAddress)
+
+	listener, err := net.Listen(networkType, listenAddr)
+	if err != nil {
+		return fmt.Errorf("listen (%s, %s) failed: %v", networkType, listenAddr, err)
+	}
+	s.listener = listener
+
+	return nil
+}
+
+// setupTLSListener sets the server TLS listener if the HTTPS server is enabled.
+// The HTTPS server can be disabled by setting the SecureBindAddress to "-" or by
+// leaving it empty.
+func (s *server) setupTLSListener(opts Opts) error {
+	if opts.SecureBindAddress == "" || opts.SecureBindAddress == "-" {
+		// No HTTPS listener required
+		return nil
+	}
+
+	config := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
+		NextProtos: []string{"http/1.1"},
+	}
+	if opts.TLS != nil {
+		cert, err := getCertificate(opts.TLS)
+		if err != nil {
+			return fmt.Errorf("could not load certificate: %v", err)
+		}
+		config.Certificates = []tls.Certificate{cert}
+	}
+
+	listener, err := net.Listen("tcp", opts.SecureBindAddress)
+	if err != nil {
+		return fmt.Errorf("listen (%s) failed: %v", opts.SecureBindAddress, err)
+	}
+
+	s.tlsListener = tls.NewListener(tcpKeepAliveListener{listener.(*net.TCPListener)}, config)
+	return nil
+}
+
+// Start starts the HTTP and HTTPS server if applicable.
+// It will block until the context is cancelled.
+// If any errors occur, only the first error will be returned.
+func (s *server) Start(ctx context.Context) error {
+	g, groupCtx := errgroup.WithContext(ctx)
+
+	if s.listener != nil {
+		g.Go(func() error {
+			if err := s.startServer(groupCtx, s.listener); err != nil {
+				return err
+			}
+			return nil
+		})
+	}
+
+	if s.tlsListener != nil {
+		g.Go(func() error {
+			if err := s.startServer(groupCtx, s.tlsListener); err != nil {
+				return err
+			}
+			return nil
+		})
+	}
+
+	return g.Wait()
+}
+
+// startServer creates and starts a new server with the given listener.
+// When the given context is cancelled the server will be shutdown.
+// If any errors occur, only the first error will be returned.
+func (s *server) startServer(ctx context.Context, listener net.Listener) error {
+	srv := &http.Server{Handler: s.handler}
+	g, groupCtx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		<-groupCtx.Done()
+
+		if err := srv.Shutdown(context.Background()); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		if err := srv.Serve(listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			return fmt.Errorf("could not start server: %v", err)
+		}
+		return nil
+	})
+
+	return g.Wait()
+}
+
+// getNetworkScheme gets the scheme for the HTTP server.
+func getNetworkScheme(addr string) string {
+	var scheme string
+	i := strings.Index(addr, "://")
+	if i > -1 {
+		scheme = addr[0:i]
+	}
+
+	switch scheme {
+	case "", "http":
+		return "tcp"
+	default:
```

Comment:
I feel like a common misconfiguration would be `https` somehow here -- Is that handled somewhere else since that wouldn't align to the `net.Listener` legal networks?

> The network must be "tcp", "tcp4", "tcp6", "unix" or "unixpacket".

Reply:
Oh this is coming straight over from our existing codebase - carry on 😄

Reply:
You raise an interesting point, we should probably investigate, but yeah, this is lift and shift so not sure 🤔

Reply:
We don't use this on the https listener, that is forced to tcp as that's the only one that makes sense, so we don't need to handle this case separately I don't think